### PR TITLE
Gamepad

### PR DESCRIPTION
**Pull request by UvuvDev**
#### Summary

A controller wrapper around the PROS Controller that adds extra functionality, such as shift keys, toggles, controller map modes, and auto-running functions.

#### Motivation

Add extra functionality to both the controller and LemLib.

#### Test Plan

Unit tests, and running it on the bot

## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/7591652397)
- via manual download: [LemLib@0.5.0-rc2+1a61cf.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1182552472.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc2+1a61cf.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1182552472.zip;
pros c fetch LemLib@0.5.0-rc2+1a61cf.zip;
pros c apply LemLib@0.5.0-rc2+1a61cf;
rm LemLib@0.5.0-rc2+1a61cf.zip;
```

### CODE REVIEW

**Comment by UvuvDev** [comment]
Also, this has been tested and it works.

**Comment by liam-teale** [comment]
has this been tested again after merging?

File: include/lemlib/devices/gamepad.hpp [hunk 1]
```diff
@@ -0,0 +1,283 @@
+#include "pros/device.hpp"
+#include "unordered_map"
+#include "pros/misc.h"
+#include <unordered_map>
+
+namespace lemlib {
+
+struct LEMButtonMapping {
```
Comment (liam-teale) [inline include/lemlib/devices/gamepad.hpp]: Not necessary to preface names with `LEM`

File: include/lemlib/devices/gamepad.hpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
+#include "pros/device.hpp"
+#include "unordered_map"
```
Comment (liam-teale) [inline include/lemlib/devices/gamepad.hpp]:
```suggestion
```
you have it included below

File: src/lemlib/movements/pursuit.cpp
Comment (liam-teale) [inline src/lemlib/movements/pursuit.cpp:1]: why has this been renamed?

File: src/lemlib/devices/gamepad.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }};
+
+    return true;
+}
+
+void Gamepad::autoButtonFunctions() {
+    // for (int i = 0; i < buttonStates.size(); i++) {
+    //     buttonStates.at(buttonsToFunctions.at(i)->getButton()) =
+    //     prosController->get_digital(buttonsToFunctions.at(i)->getButton());
+    // }
```
Comment (liam-teale) [inline src/lemlib/devices/gamepad.cpp:176]: you probably want to remove this

File: include/lemlib/devices/gamepad.hpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                case pros::E_CONTROLLER_ANALOG_LEFT_X: key = LeftXKey; break;
+                case pros::E_CONTROLLER_ANALOG_RIGHT_X: key = RightXKey; break;
+            }
+
+            return key;
+        }
+};
+
+class Gamepad {
+    private:
```
Comment (liam-teale) [inline include/lemlib/devices/gamepad.hpp]:
```suggestion
```

**Review by liam-teale — CHANGES_REQUESTED** [review]

**Pull request closed without merging**